katdal/test/test_sensordata.py

```diff
@@ -71,3 +71,29 @@ def test_numpy_object(self):
         a = np.array([b'abc', u'def', (b'xyz', u'uvw')], dtype='O')
         b = np.array(['abc', 'def', ('xyz', 'uvw')], dtype='O')
         np.testing.assert_array_equal(to_str(a), b)
+
+
+class TestSensorCache(object):
+    def setup(self):
+        self.cache = SensorCache({}, timestamps=np.arange(10.), dump_period=1.0)
+
+    def test_virtual_sensors(self):
+
+        def _check_sensor(cache, name, **kwargs):
+            """Check that virtual sensor function gets the expected parameters."""
+            assert_equal(kwargs, params)
+
+        params = {'ant': 'm000', 'param1': 'one', 'param2': 'two'}
+        template = 'Antennas/{ant}/{param1}_{param2}'
+        self.cache.virtual[template] = _check_sensor
+        self.cache.get(template.format(**params))
+        # If your parameter values contain underscores, don't use it as delimiter
+        params = {'ant': 'm000', 'param1': 'one', 'param2': 'two_three'}
+        with assert_raises(AssertionError):
+            self.cache.get(template.format(**params))
+        template = 'Antennas/{ant}/{param1}/{param2}'
+        # The updated template has not yet been added to the cache
+        with assert_raises(KeyError):
+            self.cache.get(template.format(**params))
+        self.cache.virtual[template] = _check_sensor
+        self.cache.get(template.format(**params))
```

Comment:
Looks good so far, but I can suggest two more tests:
1. Have `_check_sensor` return something, and check that this becomes the value of the virtual sensor.
2. Retrieve the sensor twice, and check that `_check_sensor` is only called the first time i.e. that the cache actually caches results.

Possibly mock would help with this (it basically gives you a callable that keeps a record of the calls made to it), but it might not be worth the extra complexity.

Reply:
Done.